### PR TITLE
G-API: Implement color-convert kernels

### PR DESCRIPTION
**Pull request by TolyaTalamanov**
### This pullrequest changes
Implement kernels
* BayerGR2RGB
* RGB2HSV
* RGB2YUV422

### CODE REVIEW

File: modules/gapi/include/opencv2/gapi/imgproc.hpp [hunk 1]
```diff
@@ -187,6 +187,26 @@ namespace imgproc {
...
+        static cv::GMatDesc outMeta(cv::GMatDesc in) {
+            return in.withType(CV_8U, 3);
+        }
+    };
+
+    G_TYPED_KERNEL(GRGB2HSV, <cv::GMat(cv::GMat)>, "org.opencv.imgproc.colorconvert.rgb2hsv") {
+        static cv::GMatDesc outMeta(cv::GMatDesc in) {
+            return in.withType(CV_8U, 3);
+        }
+    };
```
Comment (dmatveev) [inline modules/gapi/include/opencv2/gapi/imgproc.hpp]: In this case meta doesn't change, right? Both RGB and HSV are 8uC3.

File: modules/gapi/perf/common/gapi_imgproc_perf_tests_inl.hpp [hunk 2]
```diff
@@ -18,6 +18,43 @@ namespace opencv_test
...
+          b = rgb_line[i * 3 + 5];
+
+          yuv422_line[i * 2 + 2] = cv::saturate_cast<uchar>(0.615 * r - 0.51499 * g - 0.10001 * b);   // V0
+          yuv422_line[i * 2 + 3] = cv::saturate_cast<uchar>(0.299 * r + 0.587   * g + 0.114   * b);   // Y1
+      }
+  }
+
+  void convertRGB2YUV422Ref(const cv::Mat& in, cv::Mat &out)
+  {
+      if (out.size() != in.size())
```
Comment (dmatveev) [inline modules/gapi/perf/common/gapi_imgproc_perf_tests_inl.hpp]: this `if` is not necessary. Just call `.create()`

Comment (dmatveev) [reply modules/gapi/perf/common/gapi_imgproc_perf_tests_inl.hpp]: Moreover it is actually buggy to have it :)

File: modules/gapi/src/backends/cpu/gcpuimgproc.cpp [hunk 3]
```diff
@@ -276,6 +277,30 @@ GAPI_OCV_KERNEL(GCPURGB2GrayCustom, cv::gapi::imgproc::GRGB2GrayCustom)
...
+    {
+        cv::cvtColor(in, out, cv::COLOR_RGB2HSV);
+    }
+};
+
+GAPI_OCV_KERNEL(GCPURGB2YUV422, cv::gapi::imgproc::GRGB2YUV422)
+{
+    static void run(const cv::Mat& in, cv::Mat &out)
+    {
+        cv::gapi::convertRGB2YUV422(in, out);
```
Comment (dmatveev) [inline modules/gapi/src/backends/cpu/gcpuimgproc.cpp]: What is this? do we have a cv::Mat-based function in public API now?

File: modules/gapi/src/backends/cpu/gcpuimgproc.cpp [hunk 4]
```diff
@@ -303,6 +328,25 @@ cv::gapi::GKernelPackage cv::gapi::imgproc::cpu::kernels()
...
+    }
+
+    for (int i = 0; i < in.rows; ++i)
+    {
+        const uchar* in_line_p  = in.ptr<uchar>(i);
+        uchar* out_line_p = out.ptr<uchar>(i);
+        // FIXME: scalar and vectorized version are not bitwise due to round (~0.1%)
+        cv::gapi::fluid::run_rgb2yuv422_impl(out_line_p, in_line_p, in.cols);
+    }
+}
```
Comment (dmatveev) [inline modules/gapi/src/backends/cpu/gcpuimgproc.cpp]: I believe you just can put this code block into the OCV kernel body and drop this function.

File: modules/gapi/src/backends/cpu/gcpuimgproc.hpp [hunk 5]
```diff
@@ -18,6 +18,14 @@ namespace cv { namespace gimpl {
...
 void loadCPUImgProc(std::map<std::string, cv::GCPUKernel> &kmap);
 
-}}
+}
+
+namespace gapi {
+
+void convertRGB2YUV422(const cv::Mat& in, cv::Mat& out);
+
+}
```
Comment (dmatveev) [inline modules/gapi/src/backends/cpu/gcpuimgproc.hpp]: Just remove this

File: modules/gapi/perf/cpu/gapi_imgproc_perf_tests_fluid.cpp [hunk 6]
```diff
@@ -173,6 +173,21 @@ INSTANTIATE_TEST_CASE_P(YUV2BGRPerfTestFluid, YUV2BGRPerfTest,
...
+INSTANTIATE_TEST_CASE_P(RGB2YUV422PerfTestFluid, RGB2YUV422PerfTest,
+        Combine(Values(ToleranceColor(1e-3).to_compare_f()),
+            Values(szVGA, sz720p, sz1080p),
+            Values(cv::compile_args(IMGPROC_FLUID))));
+
+INSTANTIATE_TEST_CASE_P(RGB2HSVPerfTestFluid, RGB2HSVPerfTest,
+        Combine(Values(ToleranceColor(1e-3).to_compare_f()),
+            Values(szVGA, sz720p, sz1080p),
+            Values(cv::compile_args(IMGPROC_FLUID))));
+
```
Comment (dmatveev) [inline modules/gapi/perf/cpu/gapi_imgproc_perf_tests_fluid.cpp]: Is there accuracy tests for these kernels?

Comment (TolyaTalamanov) [reply modules/gapi/perf/cpu/gapi_imgproc_perf_tests_fluid.cpp]: These are perfomance tests in which the accuracy is checked.

File: modules/gapi/include/opencv2/gapi/imgproc.hpp (hunk 1 above)
Comment (TolyaTalamanov) [reply modules/gapi/include/opencv2/gapi/imgproc.hpp]: Fixed

File: modules/gapi/perf/common/gapi_imgproc_perf_tests_inl.hpp (hunk 2 above)
Comment (TolyaTalamanov) [reply modules/gapi/perf/common/gapi_imgproc_perf_tests_inl.hpp]: Fixed

Comment (TolyaTalamanov) [reply modules/gapi/perf/common/gapi_imgproc_perf_tests_inl.hpp]: Fixed

File: modules/gapi/src/backends/cpu/gcpuimgproc.cpp (hunk 3 above)
Comment (TolyaTalamanov) [reply modules/gapi/src/backends/cpu/gcpuimgproc.cpp]: Fixed

File: modules/gapi/src/backends/cpu/gcpuimgproc.cpp (hunk 4 above)
Comment (TolyaTalamanov) [reply modules/gapi/src/backends/cpu/gcpuimgproc.cpp]: Fixed

File: modules/gapi/src/backends/cpu/gcpuimgproc.hpp (hunk 5 above)
Comment (TolyaTalamanov) [reply modules/gapi/src/backends/cpu/gcpuimgproc.hpp]: Fixed

**Comment by TolyaTalamanov** [comment]
@dmatveev

**Comment by TolyaTalamanov** [comment]
@smirnov-alexey

File: modules/gapi/perf/common/gapi_imgproc_perf_tests_inl.hpp [hunk 7]
```diff
@@ -18,6 +18,40 @@ namespace opencv_test
...
+  {
+      CV_Assert(width % 2 == 0);
+
+      for (int i = 0; i < width; i += 2)
+      {
+          uchar r = rgb_line[i * 3    ];
+          uchar g = rgb_line[i * 3 + 1];
+          uchar b = rgb_line[i * 3 + 2];
+
+          yuv422_line[i * 2    ] = cv::saturate_cast<uchar>(-0.14713 * r - 0.28886 * g + 0.436 * b);  // U0
```
Comment (smirnov-alexey) [inline modules/gapi/perf/common/gapi_imgproc_perf_tests_inl.hpp]: Need to add  + 128.f inside cast

File: modules/gapi/perf/common/gapi_imgproc_perf_tests_inl.hpp [hunk 8]
```diff
@@ -18,6 +18,40 @@ namespace opencv_test
...
+          yuv422_line[i * 2 + 1] = cv::saturate_cast<uchar>( 0.299   * r + 0.587   * g + 0.114 * b);  // Y0
+
+          r = rgb_line[i * 3 + 3];
+          g = rgb_line[i * 3 + 4];
+          b = rgb_line[i * 3 + 5];
+
+          yuv422_line[i * 2 + 2] = cv::saturate_cast<uchar>(0.615 * r - 0.51499 * g - 0.10001 * b);   // V0
```
Comment (smirnov-alexey) [inline modules/gapi/perf/common/gapi_imgproc_perf_tests_inl.hpp]: Need to add  + 128.f inside cast

File: modules/gapi/src/backends/fluid/gfluidimgproc_func.simd.hpp [hunk 9]
```diff
@@ -247,6 +268,456 @@ void run_rgb2gray_impl(uchar out[], const uchar in[], int width,
...
+
+            v_store_interleave(out + w * 3 + 3, b_low, g_low, r_low);
+            v_store_interleave(out + w * 3 + vectorStep * 3 + 3, b_high, g_high, r_high);
+
+            // output offset for scalar code
+            j += vectorStep * 2;
+        }
+    #endif
+
+    // Anatoly's scalar code
```
Comment (smirnov-alexey) [inline modules/gapi/src/backends/fluid/gfluidimgproc_func.simd.hpp]: Change or remove this comment, please

File: modules/gapi/src/backends/fluid/gfluidimgproc_func.simd.hpp [hunk 10]
```diff
@@ -247,6 +268,456 @@ void run_rgb2gray_impl(uchar out[], const uchar in[], int width,
...
```
Comment (smirnov-alexey) [inline modules/gapi/src/backends/fluid/gfluidimgproc_func.simd.hpp]: Change or remove this comment, please

File: modules/gapi/perf/cpu/gapi_imgproc_perf_tests_fluid.cpp (hunk 6 above)
Comment (dmatveev) [reply modules/gapi/perf/cpu/gapi_imgproc_perf_tests_fluid.cpp]: Yes, but nobody runs performance tests to check accuracy.
The regular expectation is that once you executed `opencv_test_gapi`, you have your functionality validated.
Which tests do you run locally before opening a code review?

Comment (dmatveev) [reply modules/gapi/perf/cpu/gapi_imgproc_perf_tests_fluid.cpp]: Thanks for adding!

**Comment by dmatveev** [comment]
Seems only Alexey's stuff is remaining

**Comment by TolyaTalamanov** [comment]
@dmatveev @smirnov-alexey @rgarnov

File: modules/gapi/perf/common/gapi_imgproc_perf_tests_inl.hpp (hunk 7 above)
Comment (TolyaTalamanov) [reply modules/gapi/perf/common/gapi_imgproc_perf_tests_inl.hpp]: Fixed

File: modules/gapi/perf/common/gapi_imgproc_perf_tests_inl.hpp (hunk 8 above)
Comment (TolyaTalamanov) [reply modules/gapi/perf/common/gapi_imgproc_perf_tests_inl.hpp]: Fixed

File: modules/gapi/src/backends/fluid/gfluidimgproc_func.simd.hpp (hunk 9 above)
Comment (TolyaTalamanov) [reply modules/gapi/src/backends/fluid/gfluidimgproc_func.simd.hpp]: Fixed

File: modules/gapi/src/backends/fluid/gfluidimgproc_func.simd.hpp (hunk 10 above)
Comment (TolyaTalamanov) [reply modules/gapi/src/backends/fluid/gfluidimgproc_func.simd.hpp]: Fixed

**Comment by dmatveev** [comment]
LGTM :+1: thanks

**Comment by dmatveev** [comment]
LGTM :+1: thanks

**Comment by dmatveev** [comment]
Just get the confirmation from @smirnov-alexey

**Comment by TolyaTalamanov** [comment]
@dmatveev

**Comment by dmatveev** [comment]
Your diff seem to have some extra stuff now. Please clear it out

**Comment by TolyaTalamanov** [comment]
@dmatveev

**Comment by TolyaTalamanov** [comment]
@rgarnov @smirnov-alexey

File: modules/gapi/src/backends/fluid/gfluidimgproc.cpp [hunk 11]
```diff
@@ -1683,6 +1683,122 @@ GAPI_FLUID_KERNEL(GFluidMedianBlur, cv::gapi::imgproc::GMedianBlur, false)
...
+        }
+    }
+
+    static cv::gapi::fluid::Border getBorder(const cv::GMatDesc&)
+    {
+        int  borderType  = cv::BORDER_CONSTANT;
+        auto borderValue = cv::Scalar();
+
+        return { borderType, borderValue };
+        //return {cv::BORDER_CONSTANT, cv::gapi::own::Scalar{0, 0, 0}};
```
[the remark this answers is not in the document]
Comment (rgarnov) [reply modules/gapi/src/backends/fluid/gfluidimgproc.cpp]: (I mean this commented line)

**Pull request merged**